linq4j/src/main/java/org/apache/calcite/linq4j/tree/BinaryExpression.java

```diff
@@ -255,11 +255,40 @@ void accept(ExpressionWriter writer, int lprec, int rprec) {
     if (writer.requireParentheses(this, lprec, rprec)) {
       return;
     }
+    // When comparing two primitive boxing class (e.g., Long x, Long y),
+    // if "x" or "y" is not a constant null expression, we should
+    // call "Objects.equals(x, y)", rather than "x == y"
+    if (nodeType == ExpressionType.Equal
+        || nodeType == ExpressionType.NotEqual) {
+      Primitive boxPrimitive0 = Primitive.ofBox(expression0.getType());
+      Primitive boxPrimitive1 = Primitive.ofBox(expression1.getType());
+      if (boxPrimitive0 != null && boxPrimitive1 != null
+          && !isConstantNullExpression(expression0)
+          && !isConstantNullExpression(expression1)) {
```

Comment:
I think we should move this logic into [1]

[1] https://github.com/apache/calcite/blob/08dfd170ae91162ffb082f48c4186952955a73d2/linq4j/src/main/java/org/apache/calcite/linq4j/tree/BinaryExpression.java#L103

Reply:
Do you mean that we should keep the evaluation logic and codegen logic as same?

Reply:
No, i misunderstood, sorry for that ~